[HLSL] Apply resource attributes to the resource type rather than the handle member

hekota opened the pull request:
Converts existing resource attributes `[[hlsl::resource_class(..)]]` and `[[is_rov]]` from declaration attributes to type attributes. 

During type attribute processing all HLSL resource type attributes are validated and collected by `SemaHLSL` (`SemaHLSL::handleResourceTypeAttr`). At the end of the declaration they are be combined into a single `HLSLAttributedResourceType` instance (`SemaHLSL::ProcessResourceTypeAttributes`) that wraps the original type and stores all of the necessary information about the resource. 

`SemaHLSL` will also need to short-term-store the `TypeLoc` information for the newly created type that will be grabbed by `TypeSpecLocFiller` soon after it is created.

Updates all places that expected resource attributes on declarations like resource binding diagnostic, builtin types in HLSLExternalSemaSource, or codegen.

Also includes implementation of `TreeTransform<Derived>::TransformHLSLAttributedResourceType` that enables the use of attributed resource types inside templates.

Fixes #104861
Part 2/2

hekota committed:
[HLSL] Apply resource attributes to the resource type rather than the handle member (#6)

hekota committed:
clang-format, cleanup

hekota committed:
Update comment

damyanp commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -556,46 +562,120 @@ void SemaHLSL::handleShaderAttr(Decl *D, const ParsedAttr &AL) {
+  auto I = LocsForHLSLAttributedResources.find(RT);
+  if (I != LocsForHLSLAttributedResources.end()) {
+    SourceLocation Loc = I->second;
+    LocsForHLSLAttributedResources.erase(RT);
```
```suggestion
    LocsForHLSLAttributedResources.erase(I);
```

Doing this should mean we can avoid looking up RT in the map a second time since we already know where it is with `I`.

damyanp commented on clang/test/ParserHLSL/hlsl_resource_class_attr.hlsl:
```diff
@@ -1,32 +1,32 @@
+// CHECK: CXXRecordDecl 0x{{[0-9a-f]+}} <col:22, line:25:1> line:23:29 struct MyBuffer2 definition
+// CHECK: CXXRecordDecl 0x{{[0-9a-f]+}} <col:22, col:29> col:29 implicit struct MyBuffer2
+// CHECK: FieldDecl 0x{{[0-9a-f]+}} <line:24:3, col:35> col:35 h 'T {{\[\[}}hlsl::resource_class(UAV)]]':'T'
+template<typename T> struct MyBuffer2 {
+  T [[hlsl::resource_class(UAV)]] h;
```
Do we have special code supporting this scenario? Because I don't think it is one that we'd ever want to support.  `h` should be a handle type, whereas `T` here is strongly suggested to be the element type of the buffer.

`myBuffer2` below at least should using something more handle-type like.

damyanp replied:
There's multiple tests below (that are modifications of existing ones) that are still using this field with the element type rather than a handle type, but I won't put comments on all of them.

hekota replied:
I've updated the tests to use `__hlsl_resource_t` everywhere we use the resource attributes.

damyanp reviewed (commented):
Typo in description:

> Converts existing resource attributes [[hlsl::resource_class(..)]] and [[is_rov]] ~~to~~ from declaration attributes to type attributes.


There's many places in the existing tests where these attributes are being used in ways we don't really intend them to use.  Maybe now would be the right time to fix them up?

hekota committed:
Address code review feedback

- use __hlsl_resource_t in tests
- move function out of SemaHLSL
- call erase with iterator

hekota commented on clang/include/clang/Sema/SemaHLSL.h:
```diff
@@ -77,6 +77,23 @@ class SemaHLSL : public SemaBase {
   QualType getInoutParameterType(QualType Ty);
+
+  // FIXME: This can be hidden (as static function in SemaHLSL.cpp) once we no
+  // longer need to create builtin buffer types in HLSLExternalSemaSource.
```
I've move it out.

bogner commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12364,6 +12364,7 @@ def err_hlsl_packoffset_cross_reg_boundary : Error<"packoffset cannot cross regi
 def err_hlsl_packoffset_alignment_mismatch : Error<"packoffset at 'y' not match alignment %0 required by %1">;
 def err_hlsl_pointers_unsupported : Error<
   "%select{pointers|references}0 are unsupported in HLSL">;
+def err_missing_resource_class : Error<"HLSL resource needs to have [[hlsl::resource_class()]] attribute">;
```
This is hlsl specific, so it should be `err_hlsl_missing_resource_class`.

bogner commented on clang/lib/AST/TypePrinter.cpp:
```diff
@@ -2062,6 +2066,7 @@ void TypePrinter::printBTFTagAttributedAfter(const BTFTagAttributedType *T,
 void TypePrinter::printHLSLAttributedResourceBefore(
     const HLSLAttributedResourceType *T, raw_ostream &OS) {
   printBefore(T->getWrappedType(), OS);
+  printAfter(T->getWrappedType(), OS);
```
This looks suspicious - should `printHLSLAttributedResourceBefore` really be calling `printAfter`?

hekota replied:
Yea, we need to print the whole wrapped type before printing the attributes. I will move the attribute printing to printHLSLAttributedResourceAfter.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -556,46 +562,120 @@ void SemaHLSL::handleShaderAttr(Decl *D, const ParsedAttr &AL) {
-    return;
-  }
+bool clang::CreateHLSLAttributedResourceType(
+    Sema &S, QualType Wrapped, llvm::SmallVector<const Attr *> &AttrList,
```
We should be able to use an `ArrayRef<const Attr *> AttrList` instead of the SmallVector reference here since we just iterate over it. Note that if we did need the SmallVector API it would be better to use `SmallVectorImpl` instead since it doesn't encode a specific size.

hekota replied:
Cool! I did not know SmallVector implicitly converts to ArrayRef when the argument requires it!

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -556,46 +562,120 @@ void SemaHLSL::handleShaderAttr(Decl *D, const ParsedAttr &AL) {
-    Diag(ArgLoc, diag::warn_attribute_type_not_supported)
-        << "ResourceClass" << Identifier;
-    return;
+  bool hasResourceClass = false;
```
Style nit: `HasResourceClass`

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -556,46 +562,120 @@ void SemaHLSL::handleShaderAttr(Decl *D, const ParsedAttr &AL) {
-        << "ResourceClass" << Identifier;
-    return;
+  bool hasResourceClass = false;
+  for (auto *Attr : AttrList) {
```
`auto` doesn't save us anything here, other than making this a bit confusing by having a variable with the same name as a type. `for (const Attr *A : AttrList)` would be better.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -556,46 +562,120 @@ void SemaHLSL::handleShaderAttr(Decl *D, const ParsedAttr &AL) {
+      if (!hasResourceClass) {
+        ResAttrs.ResourceClass = RC;
+        hasResourceClass = true;
+      } else if (RC != ResAttrs.ResourceClass) {
+        S.Diag(Attr->getLocation(), diag::warn_duplicate_attribute) << Attr;
+        return false;
+      }
```
Do we really want `__hlsl_resource_t [[hlsl::resource_class(SRV)]][[hlsl::resource_class(SRV)]] H` to be valid? Maybe we should warn even if the RC does match.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -556,46 +562,120 @@ void SemaHLSL::handleShaderAttr(Decl *D, const ParsedAttr &AL) {
+                                       HLSLResourcesTypeAttrs, QT)) {
+    const HLSLAttributedResourceType *RT =
+        dyn_cast<HLSLAttributedResourceType>(QT.getTypePtr());
+    SourceLocation Loc = HLSLResourcesTypeAttrs[0]->getLoc();
```
Should we assert that the other locations match the first one?

hekota replied:
I am not sure what you mean.

bogner replied:
I mean something like
```c++
    for (int I = 1, E = HLSLResourceTypeAttrs.size(); I != E; ++I)
      assert(HLSLResourceTypeAttrs[I]->getLoc() == Loc &&
             "Attributes for resource defined at different locations?");
```

I guess I'm really trying to ask:
- Is it possible for the locations to differ between different attributes?
- If so, why is using the first one correct?

hekota replied:
Every attribute will have a different location based on where it is in the source code. I picked the location of the first one to be the location of the aggregated type because the attributes are stored in `HLSLResourceTypeAttrs` in the same order as they are parsed.

bogner replied:
Can you add a comment explaining that here please?

hekota replied:
Will do!

hekota committed:
code review feedback

hekota committed:
update warning message

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into resource-attr-on-type

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -559,46 +563,123 @@ void SemaHLSL::handleShaderAttr(Decl *D, const ParsedAttr &AL) {
+      continue;
+    switch (A->getKind()) {
+    case attr::HLSLResourceClass: {
+      llvm::dxil::ResourceClass RC =
+          dyn_cast<HLSLResourceClassAttr>(A)->getResourceClass();
```
This should use `cast` rather than `dyn_cast`, since you know it has to be correct. Generally `dyn_cast` should always be proceeded by a null check.

```suggestion
      llvm::dxil::ResourceClass RC =
          cast<HLSLResourceClassAttr>(A)->getResourceClass();
```

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -559,46 +563,123 @@ void SemaHLSL::handleShaderAttr(Decl *D, const ParsedAttr &AL) {
+    case attr::HLSLResourceClass: {
+      llvm::dxil::ResourceClass RC =
+          dyn_cast<HLSLResourceClassAttr>(A)->getResourceClass();
+      if (!HasResourceClass) {
+        ResAttrs.ResourceClass = RC;
+        HasResourceClass = true;
+      } else {
+        S.Diag(A->getLocation(), ResAttrs.ResourceClass == RC
+                                     ? diag::warn_duplicate_attribute_exact
+                                     : diag::warn_duplicate_attribute)
+            << A;
+        return false;
+      }
```
Checking the positive condition is generally more understandable than the negative. Also, since the diagnostic branch returns early you don't need the else that way:
```suggestion
      if (HasResourceClass) {
        S.Diag(A->getLocation(), ResAttrs.ResourceClass == RC
                                     ? diag::warn_duplicate_attribute_exact
                                     : diag::warn_duplicate_attribute)
            << A;
        return false;
      }
      ResAttrs.ResourceClass = RC;
      HasResourceClass = true;
```

bogner reviewed (approved):
A couple more very minor comments then this LGTM!

hekota committed:
code review feedback - add comment, rearrange code, use cast instead of dyn_cast
